napari/utils/events/dataclass.py

```diff
@@ -180,15 +188,58 @@ def set_with_events(self: C, name: str, value: Any) -> None:
 
     # if different we emit the event with new value
     after = getattr(self, name)
-    if isinstance(before, np.ndarray) or isinstance(after, np.ndarray):
-        different = np.any(before != after)
-    else:
-        different = before != after
-    if different:
+    if not self.__equality_checks__.get(name, is_equal)(after, before):
         # use gettattr again in case `_on_name_set` has modified it
         getattr(self.events, name)(value=after)  # type: ignore
 
 
+def is_equal(v1, v2):
+    """
+    Function for basic comparison compare.
+
+    Parameters
+    ----------
+    v1: Any
+        first value
+    v2: Any
+        second value
+
+    Returns
+    -------
+    result : bool
+        Returns ``True`` if ``v1`` and ``v2`` are equivalent.
+        If an exception is raised during comparison, returns ``False``.
+
+    Warns
+    -----
+    UserWarning
+        If an exception is raised during comparison.
+    """
+    try:
+        return bool(v1 == v2)
+    except Exception as e:
+        warnings.warn(
+            "Comparison method failed. Returned False. "
+            f"There may be need to define custom _on_<name>_set method. Exception {e}"
+        )
+        return False
```

Comment:
Something like this was implemented in original PR but I was requested to change its:
https://github.com/napari/napari/pull/1966#issuecomment-737449663

Currently, when types are properly declared NumPy arrays are compared using `np.array_equal` and ask arrays using `operator.is_` 

The pointed part of the code should be used only for types that could be easily compared but have a fallback mode to prevent program crashes on comparison failiture.

Reply:
> Currently, when types are properly declared NumPy arrays are compared using np.array_equal and ask arrays using operator.is_

Ah, sorry, I missed this fact, and thought we were relying on the exception catching to compare NumPy arrays. Sorry about that!